crypto/fipsmodule/cipher/e_aesccm.c

```diff
@@ -456,3 +496,241 @@ DEFINE_METHOD_FUNCTION(EVP_AEAD, EVP_aead_aes_128_ccm_matter) {
   out->seal_scatter = aead_aes_ccm_seal_scatter;
   out->open_gather = aead_aes_ccm_open_gather;
 }
+
+static int cipher_aes_ccm_init(EVP_CIPHER_CTX *ctx, const uint8_t *key,
+                        const uint8_t *iv, int enc) {
+  CIPHER_AES_CCM_CTX *cipher_ctx = CCM_CTX(ctx);
+  if (!iv && !key) {
+    return 1;
+  }
+  if (key) {
+    block128_f block;
+    ctr128_f ctr = aes_ctr_set_key(&cipher_ctx->ks.ks, NULL, &block, key,
+                                   ctx->key_len);
+    CRYPTO_ccm128_init(&cipher_ctx->ccm, block, ctr, cipher_ctx->M,
+                       cipher_ctx->L);
+    cipher_ctx->key_set = 1;
+  }
+  if (iv) {
+    CRYPTO_ccm128_init(&cipher_ctx->ccm, NULL, NULL, cipher_ctx->M,
+                       cipher_ctx->L);
+    OPENSSL_memcpy(cipher_ctx->nonce, iv, CCM_L_TO_NONCE_LEN(cipher_ctx->L));
+    cipher_ctx->iv_set = 1;
+  }
+  return 1;
+}
+
+static int cipher_aes_ccm_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out,
+                                 const uint8_t *in, size_t len) {
+  CIPHER_AES_CCM_CTX *cipher_ctx = CCM_CTX(ctx);
+  CCM128_CTX *ccm_ctx = CCM_INNER_CTX(cipher_ctx);
+  CCM128_STATE *ccm_state = CCM_INNER_STATE(cipher_ctx);
+
+  // Implicit EVP_*Final call. CCM does all the work in EVP_*Update
+  // n.b. |out| is non-NULL and |in| is NULL despite being a no-op.
+  if (in == NULL && out != NULL) {
+    return 0;
+  }
+
+  if (!cipher_ctx->iv_set || !cipher_ctx->key_set) {
+    return -1;
+  }
+
+  if (!out) {
+    if (!in) {
+      // If |out| and |in| are both NULL, |len| is the total length of the
+      // message which we need to include that in the 0th block of the CBC-MAC.
+      cipher_ctx->message_len = len;
```

Comment:
This risk implicit truncation, no? `len` has higher width than `cipher_text->message_len`.

Reply:
That's a great catch! Not only is this an implicit truncation that depends on your architecture, this is actually out of specification. We need to make sure that message_len (and also the AAD len) have a 64-bit integer width.

```
# Here's our range for message length
> The message m, consisting of a string of l(m) octets where 0 <= l(m) < 2^(8*L)

# Here's our range for L
> Valid values of L range between 2 octets and 8 octets

# Subbing that in, we get a max m(l) of
2^(8*8) = 2^64 = 16 EXABYTES

# AAD has a similar max length
> Additional authenticated data a, consisting of a string of l(a) octets where 0 <= l(a) < 2^64
```

Reply:
Looked into this more. We're not able to support the maximum message length on 32-bit machines because EVP_CIPHER's [cipher](https://github.com/awslabs/aws-lc/blob/aec07944cd70a425723461edc1543065d442b6d9/crypto/fipsmodule/cipher/internal.h#L182) interface uses size_t for the len parameter. We have logic to truncate the algorithm's maximum length to whatever fits inside a `size_t` [here](https://github.com/awslabs/aws-lc/blob/da93b7ed5aa049c691231622594400713a9dd708/crypto/fipsmodule/cipher/e_aesccm.c#L130). You're right that there's absolutely an implicit truncation bug here. For CCM modes with L>4, running on a 64-bit machine, and messages >= 2^32, we would've done something weird. The fix here is to simply change the message_len type to size_t which will match whatever len value we get from the call.